### ISSUE TITLE
`with ComputationGraph():` syntax?

### ISSUE THREAD

**Issue by pmichel31415** [clab/dynet#1246]
From my own experience with new dynet (python) users, one of the most common mistakes is forgetting to renew the computation graph (in particular at test time where it sometimes doesn't run out of memory but becomes super slow).

What do you think of adding something like:

```python
with ComputationGraph():
    # do stuff within one computation graph
# Every expression here is stale
```

I don't think it would be more work and it could easily be made backward compatible by keeping `renew_cg` and just rewriting the docs to promote the new preffered syntax.

Thoughts?

**Comment by neubig** [comment]
FWIW, I think this is a pretty good idea.

**Comment by yoavg** [comment]
I think some logic will be hard to implement with this, but if we keep renew_cg around, then sure, this should be easy to add and potentially useful.

**Comment by yoavg** [comment]
(we can also add some lightweight tracking and write an informative warning to stderr if we identify a likely problem of that sort. For example, identify calling forward over k times in the same graph with large size increments between them. And this could be turned of in cases it is an ok behavior)

**Comment by FilippoC** [comment]
> identify calling forward over k times in the same graph
Also, calling forward after backward?

I do not use the python API so I don't know if there is any equivalent of "pass value by pointer" (e.g. pickneglogsoftmax with a point to the gold index in C++), but I don't think that is the kind of function the a new user would try.

**Comment by pmichel31415** [comment]
> I think some logic will be hard to implement with this, but if we keep renew_cg around, then sure, this should be easy to add and potentially useful.

Yes I would keep `renew_cg`, the `with` syntax would basically just call it twice (once when entering the scope and once when leaving). 

So this doesn't really change the existing API, just adds syntactic sugar to make the scope fo the computation graph more explicit.

> I do not use the python API so I don't know if there is any equivalent of "pass value by pointer" (e.g. pickneglogsoftmax with a point to the gold index in C++), but I don't think that is the kind of function the a new user would try.

Yes but what might happen is calling `.value()` (which, IIRC, does an incremental forward) at different steps in the forward pass for things like scheduled sampling.
